File: gali/src/sources/dolphin/dolphin_source.py
from math import inf
from pathlib import PurePath

from gali.sources.dolphin.dolphin_game import DolphinGame
from gali.utils.explicit_config_parser import ExplicitConfigParser
from gali.sources.game_dir import GameDir
from gali.sources.emulation_source import EmulationSource
from gali.sources.file_dependent_source import FileDependentSource


class DolphinSource(EmulationSource, FileDependentSource):

    name: str
    game_class: type[DolphinGame]
    config_path: str
    
    rom_extensions: tuple[str] = (".ciso", ".iso", ".wbfs", ".gcm", ".gcz")

    def get_config(self) -> ExplicitConfigParser:
        config = ExplicitConfigParser()
        config.read(self.config_path, encoding="utf-8-sig")
        return config

    def get_rom_dirs(self, config: ExplicitConfigParser) -> tuple[GameDir]:
        rom_dirs = []
        n_dirs = config.getint(
            "General",
            "ISOPaths",
            fallback=0
        )
        deep = config.getboolean(
            "General",
            "RecursiveISOPaths",
            fallback=False
        )
        depth = inf if deep else 0
        for i in range(n_dirs):
            path = config.get("General", f"ISOPath{i}", fallback=None)
            if path is None:
                continue
            rom_dirs.append(GameDir(path, depth))
        return tuple(rom_dirs)

    def get_rom_games(self, rom_dirs: tuple[GameDir]) -> tuple[DolphinGame]:
        games = []
        for rom_dir in rom_dirs:
            rom_paths = []
            try:
                rom_paths = self.get_rom_paths(rom_dir, self.rom_extensions)
            except OSError:
                continue
            for path in rom_paths:
                name = PurePath(path).name
                game = self.game_class(
                    name=name,
                    game_path=path,
                    is_installed=True,
                )
                games.append(game)
        return tuple(games)

    def scan(self) -> tuple[DolphinGame]:
        config = self.get_config()
        rom_dirs = self.get_rom_dirs(config)
        rom_games = self.get_rom_games(rom_dirs)
        return rom_games

    def get_precondition_file_path(self):
        return self.config_path

File: src/scanners/DolphinSource.js
const fsp = require("fs/promises");
const path = require("path");

const { GameDir } = require("./GameDir.js");
const config = require("../utils/configFormats.js");

const { EmulationSource } = require("./EmulationSource.js");
const { DolphinGame } = require("../games/DolphinGame");

const USER_DIR = process.env["HOME"];

class DolphinSource extends EmulationSource {
	
	static name = "Dolphin";
	static gameClass = DolphinGame;
	
	preferCache = false;

	configPath = `${USER_DIR}/.config/dolphin-emu/Dolphin.ini`;
	romRegex = /.+\.(c?iso|wbfs|gcm|gcz)/i;

	constructor(preferCache = false) {
		super();
		this.preferCache = preferCache;
	}

	/**
	 * Get dolphin's config data
	 * @returns {config} - Dolphin's config data
	 * @private
	 */
	async _getConfig() {

		const configFileContents = await fsp.readFile(this.configPath, "utf-8");
		const configData = config.config2js(configFileContents);

		// Check "General -> ISOPaths" value to be numeric
		const nDirs = parseInt(configData["General"]["ISOPaths"]);
		if (Number.isNaN(nDirs)) {
			throw new Error("Non numeric ISOPaths value in config file");
		}

		return configData;

	}

	/**
	 * Get dolphin's cached games
	 * @returns {DolphinGame[]} - An array of found games
	 * @see https://github.com/dolphin-emu/dolphin/blob/d5b917a6c2d25926c5aa057fdaf8fce5debb3182/Source/Core/UICommon/GameFile.h#L140
	 * @private
	 */
	async _getCachedROMs() {
		// TODO Read dolphin gamelist cache
		// Cache path : $HOME/.cache/dolphin-emu/gamelist.cache
	}

	/**
	 * Get dolphin's ROM dirs from its config data
	 * @param {object} config - Dolphin's config dara
	 * @returns {GameDir} - The game dirs extracted from dolphin's config
	 * @private
	 */
	async _getROMDirs(config) {

		const dirs = [];

		// Get number of paths and options
		if (typeof config["General"] === "undefined") { return dirs; }
		const nDirs = parseInt(config["General"]["ISOPaths"]);
		const recursive = config["General"]["RecursiveISOPaths"].toString().toLowerCase() === "true";

		// Get paths
		for (let i = 0; i < nDirs; i++) {
			const dir = config["General"][`ISOPath${i}`];
			if (typeof dir === "undefined") { continue; }
			dirs.push(new GameDir(dir, recursive));
		}

		return dirs;

	}

	/**
	 * Get dolphin ROMs from the given game dirs
	 * @param {GameDir[]} dirs - The game dirs to search ROMs into
	 * @returns {DolphinGame[]} - An array of found games
	 * @private
	 */
	async _getROMGames(dirs) {
		// TODO detect games console between GameCube and Wii
		const gamePaths = await this._getROMs(dirs, this.romRegex);
		const games = [];
		for (const gamePath of gamePaths) {
			const game = new this.constructor.gameClass(path.basename(gamePath), gamePath);
			game.isInstalled = true;
			games.push(game);
		}
		return games;
	}

	/**
	 * Get all dolphin games.
	 * @param {boolean} warn - Whether to display additional warnings
	 * @returns {DolphinGame[]} - An array of found games
	 */
	async scan(warn = false) {

		// Get config
		let config;
		try {
			config = await this._getConfig();
		} catch (error) {
			if (warn){
				console.warn(`Unable to read dolphin config file : ${error}`);
			}
		}

		// Get ROM dirs
		let romDirs = [];
		if (typeof config !== "undefined") {
			try {
				romDirs = await this._getROMDirs(config);
			} catch (error) {
				if (warn){
					console.warn(`Unable to get dolphin ROM dirs : ${error}`);
				}
			}
		}

		// Get ROM games
		let romGames = [];
		if (romDirs.length > 0) {
			try {
				romGames = await this._getROMGames(romDirs);
			} catch (error) {
				if (warn){
					console.warn(`Unable to get dolphin ROMs : ${error}`);
				}
			}
		}

		return romGames;

	}

}

module.exports = {
	DolphinSource
};